jaxlib/py_values.cc

```diff
@@ -592,8 +592,19 @@ absl::StatusOr<ShardFn> MakeShardFn(nb::handle arg, ifrt::Client* client,
                                     ifrt::Device* to_device,
                                     ifrt::MemoryKind to_memory_kind,
                                     const DevicePutOptions& options) {
-  static const absl::flat_hash_map<PyObject*,
-                                   DevicePutHandler>* const handlers = [] {
+  const absl::flat_hash_map<PyObject*,
+                            DevicePutHandler>* const handlers = [] {
+
+    // static variable assignment can deadlock under free-threading
+    // use nb mutex and lock to prevent it:
+    static nb::ft_mutex mu;
+    static absl::flat_hash_map<PyObject*, DevicePutHandler>* output = nullptr;
+    {
+      nb::ft_lock_guard lock(mu);
```

Comment:
You need to use an absl::Mutex here because this lock should apply even in a GIL-enabled build.